### PR TITLE
fix: migrate tests to Node.js test runner

### PR DESCRIPTION
**Pull request by MLopezJ**
## Changes

- Remove jest
- Migrate tests to Node.js test runner
- Improve the readability of  tests description

### CODE REVIEW

**Comment by MLopezJ** [comment]
## TODO

- [x] Improve test readability
- [x] Read about different assert types and check if correct one is used in the test depending of the context 


readability issues example:
```
convertToObject()
  ✔ should convert [object Object] of 5:1.1@1.1 to [object Object] 
```

File: package.json [hunk 1]
```diff
@@ -74,7 +57,7 @@
     ]
   },
   "engines": {
-    "node": ">=18",
+    "node": ">=20.0.0",
```
Comment (coderbyheart) [inline package.json:60]: `actions/setup-node@v4` in Actions workflows need to be adapted accordingly.

File: src/convert.spec.ts [hunk 2]
```diff
@@ -67,57 +69,62 @@ describe('convert()', () => {
...
-	it('should throw an exception if the input cannot be converted', () =>
-		expect(() =>
+	void it('should throw an exception if the input cannot be converted', () => {
+		const error = () =>
 			convert({
 				foo: {
 					'0': {
 						'Application Type': '',
 					},
 				},
```
Comment (coderbyheart) [inline src/convert.spec.ts]:
```suggestion
```

File: src/convert.spec.ts [hunk 3]
```diff
@@ -67,57 +69,62 @@ describe('convert()', () => {
...
-			}),
-		).toThrow())
+			})
+		assert.throws(error)
```
Comment (coderbyheart) [inline src/convert.spec.ts]:
```suggestion
		assert.throws(() =>
			convert({
				foo: {
					'0': {
						'Application Type': '',
					},
				},
			}))
```

This removes the not so helpful naming of the variable.

File: src/custom-conversion/validateConfig.spec.ts [hunk 4]
```diff
@@ -1,10 +1,14 @@
+import { describe, it } from 'node:test'
+import assert from 'node:assert/strict'
 import { validateConfig } from './validateConfig.js'
 
-describe('validateConfig()', () => {
-	it('mark invalid config', () =>
-		expect(
+void describe('validateConfig()', () => {
+	void it('mark invalid config', () => {
+		assert.strictEqual(
```
Comment (coderbyheart) [inline src/custom-conversion/validateConfig.spec.ts]:
```suggestion
		assert.equal(
```

Strict is implied through `'node:assert/strict'`

File: src/custom-conversion/validateCustomObjects.spec.ts [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
+	void describe('should remove invalid custom objects', () => {
+		void it('remove invalid config', () => {
 			const input = {
 				'50009@1.2': {
 					'42': '666',
 				},
 			}
 			const output = validateCustomObjects(input as any)
-			expect(output['50009@1.2']).toBeUndefined()
+			assert.deepEqual(output['50009@1.2'], undefined)
```
Comment (coderbyheart) [inline src/custom-conversion/validateCustomObjects.spec.ts]:
```suggestion
			assert.equal(output['50009@1.2'], undefined)
```

No deep equality needed.

File: src/resourceNameToUrn.spec.ts [hunk 6]
```diff
@@ -23,12 +25,12 @@ describe('resourceNameToUrn()', () => {
...
+	] as [string, unknown][]) {
+		void it(`should convert the resource name "${resourceName}" to the URN "${expectedURN}"`, () => {
+			assert.strictEqual(resourceNameToUrn(resourceName), expectedURN)
+		})
+	}
 
-	it('should return null for unknown objects', () =>
-		expect(resourceNameToUrn('foo')).toBeNull())
+	void it('should return null for unknown objects'),
+		() => assert.strictEqual(resourceNameToUrn('foo'), null)
```
Comment (coderbyheart) [inline src/resourceNameToUrn.spec.ts]:
```suggestion
		() => assert.equal(resourceNameToUrn('foo'), null)
```

File: src/resourceNameToUrn.spec.ts [hunk 7]
```diff
@@ -23,12 +25,12 @@ describe('resourceNameToUrn()', () => {
...
 		['Location Assistance', null],
-	])(
-		`should convert the resource name "%s" to the URN "%s"`,
-		(resourceName, expectedURN) =>
-			expect(resourceNameToUrn(resourceName)).toEqual(expectedURN),
-	)
+	] as [string, unknown][]) {
+		void it(`should convert the resource name "${resourceName}" to the URN "${expectedURN}"`, () => {
+			assert.strictEqual(resourceNameToUrn(resourceName), expectedURN)
+		})
```
Comment (coderbyheart) [inline src/resourceNameToUrn.spec.ts]:
```suggestion
		void it(`should convert the resource name "${resourceName}" to the URN "${expectedURN}"`, () => assert.equal(resourceNameToUrn(resourceName), expectedURN))
```

File: src/schema-conversion/convertResource.spec.ts [hunk 8]
```diff
@@ -1,212 +1,216 @@
...
-			).toEqual(expected)
-		},
-	)
+	] as [string, CoioteLwM2MObjectPropertyValue, unknown][]) {
+		void it(`should convert prop '${propId}' from object (LwM2MServer). Value: '${value}' Expected: '${expected}'`, () => {
+			assert.deepEqual(
+				convertResource(LwM2MServer as any, propId, value), // TODO: use real schema definition from lwm2m type lib
+				expected,
+			)
+		})
```
Comment (coderbyheart) [inline src/schema-conversion/convertResource.spec.ts]:
```suggestion
		)
```

File: src/schema-conversion/convertResource.spec.ts [hunk 9]
```diff
@@ -1,212 +1,216 @@
...
-				},
-				required: ['0', '1', '6', '7'],
-			}
-			expect(
-				convertResource(LwM2MServerSchema as any, propId, value as any),
-			).toEqual(expected)
-		},
-	)
+	] as [string, CoioteLwM2MObjectPropertyValue, unknown][]) {
+		void it(`should convert prop '${propId}' from object (LwM2MServer). Value: '${value}' Expected: '${expected}'`, () => {
```
Comment (coderbyheart) [inline src/schema-conversion/convertResource.spec.ts]:
```suggestion
		void it(`should convert prop '${propId}' from object (LwM2MServer). Value: '${value}' Expected: '${expected}'`, () => 
```

File: src/schema-conversion/convertResource.spec.ts [hunk 10]
```diff
@@ -216,9 +220,13 @@ describe('convertValue', () => {
...
-			expect(convertValue(value, type)).toEqual(expected),
-	)
+	] as [
+		string,
+		'integer' | 'number' | 'boolean' | 'string' | 'timestamp',
+		unknown,
+	][]) {
+		void it(`should convert value '${value}' which is type '${type}' to '${expected}'`, () => {
+			assert.deepEqual(convertValue(value, type), expected)
+		})
```
Comment (coderbyheart) [inline src/schema-conversion/convertResource.spec.ts]:
```suggestion
		void it(`should convert value '${value}' which is type '${type}' to '${expected}'`, () => assert.equal(convertValue(value, type), expected))
```

No deep equality needed.

File: src/schema-conversion/convertToArray.spec.ts [hunk 11]
```diff
@@ -149,20 +156,18 @@ describe('convertToArray()', () => {
...
-				),
-			).toMatchObject(expected),
-	)
+	] as [CoioteLwM2MObject, string, unknown][]) {
+		void it(`should convert object definition of ${urn} to expected format`, () => {
+			assert.deepEqual(
+				convertToArray(LwM2MDocumentSchema.properties[urn], objectDefinition),
+				expected,
+			)
+		})
```
Comment (coderbyheart) [inline src/schema-conversion/convertToArray.spec.ts]:
```suggestion
		void it(`should convert object definition of ${urn} to expected format`, () => 
			assert.deepEqual(
				convertToArray(LwM2MDocumentSchema.properties[urn], objectDefinition),
				expected,
			)
		)
```

File: src/schema-conversion/convertToArray.spec.ts [hunk 12]
```diff
@@ -175,15 +180,10 @@ describe('convertElementOfArray()', () => {
...
-			expect(
-				convertElementOfArray(
-					LwM2MDocumentSchema.properties[urn as any],
-					value as any,
-				),
-			).toMatchObject(expected)
-		},
-	)
+	] as [string, CoioteLwM2MObjectProperties, unknown][]) {
+		assert.deepEqual(
```
Comment (coderbyheart) [inline src/schema-conversion/convertToArray.spec.ts]: `it` is missing.

File: src/schema-conversion/convertToObject.spec.ts [hunk 13]
```diff
@@ -51,16 +61,25 @@ describe('convertToObject()', () => {
...
-			),
-		).toMatchObject(expected),
-	)
+	] as [CoioteLwM2MObject, string, unknown][]) {
+		void it(`should convert object ${urn} to expected format`, () => {
+			assert.deepEqual(
+				convertToObject(LwM2MDocumentSchema.properties[urn], objectDefinition),
+				expected,
+			)
+		})
```
Comment (coderbyheart) [inline src/schema-conversion/convertToObject.spec.ts]:
```suggestion
		void it(`should convert object ${urn} to expected format`, () => 
			assert.deepEqual(
				convertToObject(LwM2MDocumentSchema.properties[urn], objectDefinition),
				expected,
			)
		)
```

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (MLopezJ) [reply package.json:60]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/734f81605a2a2988916cb48ee8051352633667f5

File: src/convert.spec.ts (hunk 3 above)
Comment (MLopezJ) [reply src/convert.spec.ts]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/7ca94bee1cb6159a7ae9165f3bba09d5c204f0e0

File: src/convert.spec.ts (hunk 2 above)
Comment (MLopezJ) [reply src/convert.spec.ts]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/7ca94bee1cb6159a7ae9165f3bba09d5c204f0e0

File: src/schema-conversion/convertResource.spec.ts [hunk 14]
```diff
@@ -1,212 +1,214 @@
...
-			}
-			expect(
-				convertResource(LwM2MServerSchema as any, propId, value as any),
-			).toEqual(expected)
-		},
-	)
+	] as [string, CoioteLwM2MObjectPropertyValue, unknown][])
+		void it(`should convert prop '${propId}' from object (LwM2MServer). Value: '${value}' Expected: '${expected}'`, () =>
+			assert.deepEqual(
+				convertResource(LwM2MServer as any, propId, value), // TODO: use real schema definition from lwm2m type lib
```
Comment (MLopezJ) [inline src/schema-conversion/convertResource.spec.ts:205]: TODO is solved here: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/86

File: src/resourceNameToUrn.spec.ts [hunk 15]
```diff
@@ -1,7 +1,9 @@
 import { resourceNameToUrn } from './resourceNameToUrn.js'
+import { describe, it } from 'node:test'
+import assert from 'node:assert/strict'
 
-describe('resourceNameToUrn()', () => {
-	it.each([
+void describe('resourceNameToUrn', () => {
```
Comment (coderbyheart) [inline src/resourceNameToUrn.spec.ts]:
```suggestion
void describe('resourceNameToUrn()', () => {
```

We always use function name in describe.

File: src/schema-conversion/convertObjectUsingSchema.spec.ts [hunk 16]
```diff
@@ -96,14 +100,14 @@ describe('convertObjectUsingSchema()', () => {
 				{ '0': 195, '1': 0, '2': 300, '3': -16, '4': -119, '5': 23 },
 			],
 		],
-	])(
-		'should convert object definition %j for object %s using json schema to %j',
-		(objectDefinition, urn, expected) =>
-			expect(
+	] as [CoioteLwM2MObject, string, unknown][])
+		void it(`should convert object definition of object '${urn}' using json schema`, () => {
```
Comment (coderbyheart) [inline src/schema-conversion/convertObjectUsingSchema.spec.ts]:
```suggestion
		void it(`should convert object definition of object '${urn}' using json schema`, () => 
```

File: src/schema-conversion/convertObjectUsingSchema.spec.ts [hunk 17]
```diff
@@ -96,14 +100,14 @@ describe('convertObjectUsingSchema()', () => {
...
 				convertObjectUsingSchema(
 					LwM2MDocumentSchema.properties[urn],
-					objectDefinition as any,
+					objectDefinition,
 				),
-			).toMatchObject(expected),
-	)
+				expected,
+			)
+		})
```
Comment (coderbyheart) [inline src/schema-conversion/convertObjectUsingSchema.spec.ts]:
```suggestion
		)
```

File: src/schema-conversion/convertResource.spec.ts [hunk 18]
```diff
@@ -1,212 +1,214 @@
...
+				expected,
+			))
 })
 
-describe('convertValue', () => {
-	it.each([
+void describe('convertValue', () => {
```
Comment (coderbyheart) [inline src/schema-conversion/convertResource.spec.ts:210]:
```suggestion
void describe('convertValue()', () => {
```

File: src/schema-conversion/convertResource.spec.ts [hunk 19]
```diff
@@ -1,212 +1,214 @@
+import { describe, it } from 'node:test'
+import assert from 'node:assert/strict'
 import { convertResource, convertValue } from './convertResource.js'
+import type { CoioteLwM2MObjectPropertyValue } from 'src/types.js'
 
-describe('convertResource', () => {
-	it.each([
+void describe('convertResource', () => {
```
Comment (coderbyheart) [inline src/schema-conversion/convertResource.spec.ts]:
```suggestion
void describe('convertResource()', () => {
```

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: src/schema-conversion/convertResource.spec.ts (hunk 19 above)
Comment (MLopezJ) [reply src/schema-conversion/convertResource.spec.ts]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/5d82ae548dac757c132f49a0441b955f1995f46c

File: src/schema-conversion/convertResource.spec.ts (hunk 18 above)
Comment (MLopezJ) [reply src/schema-conversion/convertResource.spec.ts:210]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/5d82ae548dac757c132f49a0441b955f1995f46c

File: src/resourceNameToUrn.spec.ts (hunk 15 above)
Comment (MLopezJ) [reply src/resourceNameToUrn.spec.ts]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/5d82ae548dac757c132f49a0441b955f1995f46c

File: src/schema-conversion/convertObjectUsingSchema.spec.ts (hunk 16 above)
Comment (MLopezJ) [reply src/schema-conversion/convertObjectUsingSchema.spec.ts]: https://github.com/NordicSemiconductor/asset-tracker-cloud-coiote-aws-converter-js/pull/85/commits/98067375abb58658ec7bb9371d6b2d0d05867fd1

**Pull request merged**